openupgrade_scripts/scripts/purchase_stock/16.0.1.2/pre-migration.py

```diff
@@ -0,0 +1,73 @@
+from openupgradelib import openupgrade
+
+
+def _compute_purchase_order_receipt_status(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE purchase_order
+        ADD COLUMN IF NOT EXISTS receipt_status character varying
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        WITH po_receipt_status as (
+            SELECT
+                rel.purchase_order_id as id,
+                CASE
+                    WHEN COUNT(sp.state)
+                        FILTER(WHERE sp.state = 'cancel') = COUNT(sp.state)
+                        THEN null
+                    WHEN COUNT(sp.state)
+                        FILTER(WHERE sp.state not in ('done', 'cancel')) = 0
+                        THEN 'full'
+                    WHEN COUNT(sp.state)
+                        FILTER(WHERE sp.state = 'done') > 0
+                        THEN 'partial'
+                    ELSE 'pending'
+                END as receipt_status
+            FROM stock_picking as sp
+            JOIN purchase_order_stock_picking_rel as rel
+            ON rel.stock_picking_id = sp.id
+            GROUP BY 1
+            ORDER BY 1
+        )
+        UPDATE purchase_order as po
+        SET receipt_status = po_receipt_status.receipt_status
+        FROM po_receipt_status
+        WHERE po_receipt_status.id = po.id;
+        """,
+    )
+
+
+def _compute_purchase_order_dest_address_id(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE purchase_order
+        ADD COLUMN IF NOT EXISTS dest_address_id integer
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE purchase_order AS po
+        SET dest_address_id = FALSE
+        FROM stock_picking_type AS spt
+        JOIN stock_location AS sl
+        ON spt.default_location_dest_id = sl.id
+        WHERE po.picking_type_id = spt.id AND sl.usage != 'customer'
+        """,
+    )
```

Comment:
[the compute method](https://github.com/OCA/OCB/blob/16.0/addons/purchase_stock/models/purchase.py#L77) sets the field to `False` (thus `null`) when these conditions are true. since its a new column, all values are already `null`. i ran the migration without calling this function, and the compute function is not called (i actually don’t know exactly why; i first thought that it was because the field was writable, but even when removing `readonly=False` from the field, it is not called). anyway, nothing needs to be done (even not add the field).